keps/sig-storage/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,681 @@
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Currently, the PVCs created automatically by the StatefulSet are not deleted when 
+the StatefulSet is deleted. As can be seen by the discussion in the issue 
+[55045](https://github.com/kubernetes/kubernetes/issues/55045) there are several use
+cases where the PVCs which are automatically created are deleted as well. There is
+merit to the suggestion that the PVCs should not be deleted since they have a different
+lifecycle than the pods, hence this proposal intends to keep this as an opt in 
+for users.
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+Provide a feature to auto delete the PVCs created by StatefulSet.
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+This proposal does not plan to address how the underlying PVs are treated on PVC deletion. 
+That functionality will continue to be governed by the ReclaimPolicy of the storage class. 
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+### Background
+
+Controller `garbagecollector` is responsible for ensuring that when a statefulset 
+set is deleted the corresponding pods spawned from the StatefulSet is deleted. 
+The `garbagecollector` uses `OwnerReference` added to the `Pod` by statefulset controller
+to delete the Pod. Similar mechanism is leveraged by this proposal to automatically 
+delete the PVCs created by the StatefulSet controller.
+
+### Changes required
+
+The following changes are required:
+
+1. Add `VolumeReclaimPolicy` entry into StatefulSet spec inorder to make this feature an opt-in.
+2. Provide the following VolumeReclaimPolicies:
+   * Retain - this is the default policy and is considered in cases where no policy is specified. 
+       This would be the existing behaviour - when a StatefulSet is deleted, no action is taken with
+       respect to the PVCs created by the StatefulSet.
+   * RemoveOnPodDeletion - Whenever a pod is deleted, the corresponding PVC is deleted as well. 
```

Comment:
`RemoveOnPodDeletion` is very similar to "generic ephemeral inline volumes" (https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes): volumes get created for a pod and deleted together with it. Just the mechanism is different.

I'm undecided whether it makes sense to have two different approaches for the same use case. One the one hand, it fits into the proposed API. On the other hand, additional code will be needed to implement it and that code will be more complex than the one for `RemoveOnStatefulSetDeletion` (like dealing with slow PVC removal).

Reply:
Does that mean we want to change the semantics do that the PVC is only removed on StatefulSet scaledown? (and change the name to RemoveOnScaledown or similar)

Reply:
Using a string for the reclaim policy is better than a boolean, I would just remove the `RemoveOnPodDeletion` option unless there's a good reason for having it.

Reply:
I mean change the semantics so that the three options are:
* Retain (current behavior)
* RemoveOnScaledown (delete PVs on scaledown but not if pod is deleted/restarted for another reason such as being unhealthy)
* RemoveOnStatefulSetDeletion (as currently described)

This would fit the use cases of customer feedback that we've heard.

Reply:
Thanks @pohly @mattcary . Will update the PR.

Reply:
ok so we are saying, this KEP is mainly for the scale down scenario and not any pod deletion scenario? It would help to make this clear here

Reply:
I think that makes sense. The only options will be Retain, RemoveOnScaledown, RemoveOnStatefulSetDeletion.